File: [23.12]Design-Pattern/Java/src/creational/dynamicFactory/DynamicFactory.java
package creational.dynamicFactory;

import creational.enumFactoryMethod.Drink;
import creational.enumFactoryMethod.Food;
import creational.enumFactoryMethod.Hamburger;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Map;

public class DynamicFactory {
    // 클래스를 넣을 Map
    private static final Map<String, Class<? extends Food>> registerTypes = new HashMap<>();

    // map에 기본적으로 들어가는 클래스
    static {
        registerTypes.put("Hamburger", Hamburger.class);
        registerTypes.put("Drink", Drink.class);
    }

    // 이후 개발 도중 추가해야 되는 클래스가 생긴 경우 사용
    public static void setRegisterTypes(String type, Class<? extends Food> cls){
        registerTypes.put(type, cls);
    }

    private static Food getFood(String type) throws NoSuchMethodException, InvocationTargetException, InstantiationException, IllegalAccessException {
        // 해당 타입의 클래스 가져오기
        Class<?> cls = registerTypes.get(type);

        if(cls == null){
            throw new RuntimeException();
        }

        // 해당 클래스에서 생성자 가져오기
        Constructor<?> foodConstructor = cls.getDeclaredConstructor();

        // Reflection API를 통해 인스턴스 만들고 업캐스팅
        return (Food) foodConstructor.newInstance();
    }

    public static Food createFood(String type){
        Food food = null;

        try {
            food = getFood(type);
        } catch (NoSuchMethodException | InvocationTargetException | InstantiationException | IllegalAccessException | RuntimeException e) {
            System.err.println("해당 음식이 없습니다.");
        }
        return food;
    }

}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/creational/dynamicFactory/DynamicFactory.kt
package creational.dynamicFactory

import creational.enumFactoryMethod.Drink
import creational.enumFactoryMethod.Food
import creational.enumFactoryMethod.Hamburger
import java.lang.RuntimeException

object DynamicFactory {

    // out을 통해 하위 객체도 저장될 수 있도록 한다.
    private val registerTypes: MutableMap<String, Class<out Food>> = HashMap();

    // 기본 클래스 타입 저장
    init {
        creational.dynamicFactory.DynamicFactory.registerTypes["Hamburger"] = Hamburger::class.java
        creational.dynamicFactory.DynamicFactory.registerTypes["Drink"] = Drink::class.java
    }

    // 실행 도중 추가하고 싶어진 경우
    fun setRegisterType(type: String, cls: Class<out Food>){
        creational.dynamicFactory.DynamicFactory.registerTypes[type] = cls
    }

    // 클래스 확인 후 리턴
    private fun getFoodClass(type: String): Class<out Food> {
        return creational.dynamicFactory.DynamicFactory.registerTypes[type] ?: throw RuntimeException("해당 음식 없음")
    }

    // 클래스 생성자를 이용하여 새로운 객체를 만들고 Food로 형변환
    fun createFood(type: String): Food {
        return creational.dynamicFactory.DynamicFactory.getFoodClass(type).getDeclaredConstructor().newInstance() as Food
    }

}